### ISSUE TITLE
(core): Ability to set Stack termination protection on Stage level (and inherit by Stacks within it)

### ISSUE THREAD

**Issue by aripalo** [aws/aws-cdk#31449]
### Describe the feature

Add ability to define `Stack` **termination protection** via (parent) `Stage`, as demonstrated by an example test scenario:

```ts
test('Stack inherits termination protection from Stage', () => {
    // GIVEN
    const app = new App();
    const stage = new Stage(app, 'Stage', {
      terminationProtection: true,
    });

    // WHEN
    const stack1 = new Stack(stage, 'Stack1');

    // THEN
    expect(stack1.terminationProtection).toEqual(true);
});
```

Additionally any inner `Stage` and/or `Stack` within the (parent) Stage scope can override the inherited value.

### Use Case

Stages are a mechanism to "group" Stacks together. There are multiple use cases / reasons to group stacks using Stages, but one of them is to define an "application environment" (not to be confused with `cdk.Environment`) via a Stage: Think of `development` or `production` Stage etc. 

The specific Stage use case / reason, doesn't matter that much, but often it would be handy to be able to define Termination Protection for all Stacks within the scope of specific Stage.

One could make the statement, that this is achievable via Aspects, but as stacks default to having `terminationProtection: false`, there is no reasonable way (at least AFAIK) to detect on a stack level if termination protection is set to `false` explicitly through props or implicitly by the stack constructor. Hence the idea of a feature where one can define (the default) termination protection on Stage level, which is then inherited by all its child Stages (if any) and finally by the stacks within that Stage scope; Additionally one should be able to override the Stage-level (inherited) termination protection on Stack level.

### Proposed Solution

1. Introduce optional `terminationProtection` into `StageProps`
2. In `Stage` constructor, either use the termination protection value from its props (if set) or inherit from parent stage
3. In `Stack` constructor, either use the termination protection value from its props (if set), or inherit from parent stage, or finally default to `false` (as before)

Basically, allow this:
```ts
new Stage(app, 'Stage', {
  terminationProtection: true,
});
```

### Other Information

I added an integ-test for this feature in `packages/@aws-cdk-testing/framework-integ/test/core/test/integ.stage-termination-protection.ts` in my PR but I was not able to run it due to some unrelated issue in my dev environment.

I have an upcoming PR about this.

I think this feature counts as a "non-breaking new feature", as this has not been possible before.

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

v2.158.0

### Environment details (OS name and version, etc.)

macOS 14.6.1

**Comment by pahud** [comment]
It seems to be a very handy feature. Thank you for your PR.

**Comment by aripalo** [comment]
I think this is **very useful feature** to have and does NOT introduce any breaking changes.

I made a new PR #33239 which passes all the checks and AWS CodeBuild CI quality gates. Hoping for this getting merged 🤞 

If there's anything at all that this PR still requires to meet the criteria for a merge, let me know!

**Comment by SupermanAdam007** [comment]
Cool please merge. 👍

**Comment by aripalo** [comment]
After some discussion in the cdk.dev Slack, I verified that setting a termination protection via an Aspect applied to a Stage works too: 
![Image](https://github.com/user-attachments/assets/998ccaa1-129f-4660-8e58-3416fdc463c8)

… so that is definitely a workaround, BUT:
1. One would have to either:
    - define that aspect on every project themselves
    - or at least import & configure a dependency (of that aspect)
2. The proposed solution (IMO) is much nicer, just a one-liner, that benefits every CDK user "out-of-the-box"
3. As `cdk.Stack`s already inherit the `env` (account & region) from `cdk.Stage`, the addition of `terminationProtection` (IMO) feels like a natural addition
4. … especially as it involves relatively minimal code changes to cdk core without any breaking changes
